nemo/core/neural_factory.py

```diff
@@ -394,6 +427,14 @@ def torch_broadcast_wrapper(str_len=None, string=None, src=0):
                     return return_string
 
                 broadcast_func = torch_broadcast_wrapper
+
+                if self._model_parallel_size is not None:
+                    from megatron import mpu
+                    mpu.initialize.initialize_model_parallel(self._model_parallel_size)
+                    self._mp_rank = mpu.get_model_parallel_rank()
```

Comment:
Megatron init wouldn't work from MegatronNM?

Reply:
It would work but the idea of this PR is to initialize torch.distributed groups with the factory and then store world information like local_rank, global_rank, mp_rank, dp_rank as properties of the factory.  Note this already being done for data parallel. This has the benefit of giving easier access to world information to NeMo users.

To me this is also an advantage of NeMo's abstraction. If we build up the model parallel API enough, users could potentially use NeMo as a framework for model parallel architecture development. 

If we just call mpu.initialize_megatron from the neural module, the communication information is stored as global variables and somewhat hidden from NeMo users.

Reply:
After some more investigation, I don't think we can initialize megatron from the megatron neural module when model parallel size is greater than one. This is because the training loop is instantiated (ptActions) when the factory is initialized. In particular, DDP and DistributedSampler arguments depend on the data parallel groups.